### PR TITLE
Return (response, error) #5

### PR DESCRIPTION
**Pull request by tamird**
This is the final one before `Send`! Might be easier commit-by-commit. Work toward #1403.

### CODE REVIEW

File: kv/dist_sender.go [hunk 1]
```diff
@@ -467,13 +468,10 @@ func (ds *DistSender) sendRPC(trace *tracer.Trace, raftID proto.RaftID, replicas
...
-		// Set content of replies[0] back to reply
-		dst := reflect.ValueOf(reply).Elem()
-		dst.Set(reflect.ValueOf(replies[0]).Elem())
+	if err != nil {
+		return nil, err
 	}
-
-	return err
+	return replies[0].(proto.Response), nil
 }
```
Comment (tschottdorf) [inline kv/dist_sender.go]: what was that old code about?

Comment (tschottdorf) [inline kv/dist_sender.go]: oh, I see.

Comment (tamird) [inline kv/dist_sender.go]: the reflection stuff? it's still here, just moved to `Send`. It's basically doing `*reply = newReply`. Hurray generics.

File: kv/dist_sender.go [hunk 2]
```diff
@@ -629,7 +628,8 @@ func (ds *DistSender) Send(ctx context.Context, call proto.Call) {
 				}
 				break
 			}
-			err = ds.sendAttempt(trace, args, curReply, desc)
+			// at this point reply.Header().Error may be non-nil!
```
Comment (tschottdorf) [inline kv/dist_sender.go]: s/at/At/

File: kv/dist_sender.go [hunk 3]
```diff
@@ -616,6 +612,9 @@ func (ds *DistSender) Send(ctx context.Context, call proto.Call) {
 			// error handling below may clear them on certain errors, so we
 			// refresh (likely from the cache) on every retry.
 			descDone := trace.Epoch("meta descriptor lookup")
+			// it is safe to pass call here (with its embedded reply) because
```
Comment (tschottdorf) [inline kv/dist_sender.go]: s/it/It/

File: kv/dist_sender.go [hunk 4]
```diff
@@ -708,11 +708,15 @@ func (ds *DistSender) Send(ctx context.Context, call proto.Call) {
 			return
 		}
 
-		if finalReply != curReply {
+		if first {
+			// Modify at the pointer's target
```
Comment (tschottdorf) [inline kv/dist_sender.go]: s/$/./

**Comment by tschottdorf** [comment]
LGTM

File: kv/dist_sender.go (hunk 2 above)
Comment (tamird) [inline kv/dist_sender.go]: done

File: kv/dist_sender.go (hunk 3 above)
Comment (tamird) [inline kv/dist_sender.go]: done

File: kv/dist_sender.go (hunk 4 above)
Comment (tamird) [inline kv/dist_sender.go]: rewrote this comment

**Pull request merged**